### ISSUE TITLE
Dev code not being stripped in prod builds

### ISSUE THREAD

**Issue by vaindil** [angular/angular-cli#18603]
# 🐞 Bug report

### Command (mark with an `x`)
- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?
Problem occurs in at least Angular 9 and 10.

### Description
This issue is a follow-up to [this Stack Overflow question](https://stackoverflow.com/questions/63415056/angular-tree-shaking-not-stripping-dev-code-what-things-should-i-look-for). The question received some upvotes but no answers to the problem, so I'm asking here because I'm really not sure what to do.

The short version is that code gated by compile-time constants like below is not being stripped in prod builds in my project, and I have no idea what to look at to figure out why. I don't know if it's a bug in Angular or a problem in my code. How can I figure out the issue? Any help that can point me in the right direction would be tremendously appreciated.

I'm using this code in `main.ts` (just for testing, the actual code I need stripped is in a component):

```ts
if (environment.production) {
  console.warn('this is a prod build');
  enableProdMode();
}

if (!environment.production) {
  console.warn('this is a dev build');
}
```

And this code is output as the result of the prod build:

```js
s.X.production&&(console.warn("this is a prod build"),Object(i.R)()),s.X.production||console.warn("this is a dev build")
```

Other code:

<details><summary><code>environment.prod.ts</code></summary>
<p>

`environment.ts` is exactly the same, just with `false` instead of `true`.
```ts
export const environment = {
  production: true
};

export * from './services/services';
```
</p>
</details>

<details><summary>Relevant part of <code>angular.json</code></summary>
<p>

```json
"my-prod-config": {
  "optimization": true,
  "outputHashing": "all",
  "sourceMap": false,
  "extractCss": true,
  "namedChunks": false,
  "aot": true,
  "extractLicenses": true,
  "vendorChunk": false,
  "buildOptimizer": true,
  "stylePreprocessorOptions": {
    "includePaths": [
      "src/styles"
    ]
  },
  "fileReplacements": [
    {
      "replace": "src/environments/environment.ts",
      "with": "src/environments/environment.prod.ts"
    }
  ],
  "baseHref": "./"
}
```
</p>
</details>

<details><summary><code>tsconfig.base.json</code></summary>
<p>

```json
{
  "compileOnSave": false,
  "compilerOptions": {
    "downlevelIteration": true,
    "importHelpers": true,
    "module": "es2020",
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "baseUrl": "src/",
    "experimentalDecorators": true,
    "allowJs": true,
    "target": "es2015",
    "lib": [
      "es2018",
      "dom"
    ],
    "paths": {
      "path1": [
        "app/modules/stripped-from-stack-overflow-example1"
      ],
      "path2": [
        "app/modules/stripped-from-stack-overflow-example2"
      ]
    }
  },
  "files": [
    "src/main.ts",
    "src/polyfills.ts"
  ],
  "angularCompilerOptions": {
    "fullTemplateTypeCheck": true,
    "strictTemplates": true,
    "strictInjectionParameters": true
  }
}
```
</p>
</details>

<details><summary><code>package.json</code></summary>
<p>

```json
{
  "name": "my-app",
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {
    "section stripped": "section stripped"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "10.0.8",
    "@angular/common": "10.0.8",
    "@angular/compiler": "10.0.8",
    "@angular/core": "10.0.8",
    "@angular/forms": "10.0.8",
    "@angular/platform-browser": "10.0.8",
    "@angular/platform-browser-dynamic": "10.0.8",
    "@angular/router": "10.0.8",
    "@ng-idle/core": "9.0.0-beta.1",
    "@ng-idle/keepalive": "9.0.0-beta.1",
    "@ngneat/until-destroy": "8.0.1",
    "angular-svg-icon": "10.0.0",
    "brace": "0.11.1",
    "caniuse-lite": "1.0.30001111",
    "core-js": "3.6.5",
    "css-vars-ponyfill": "2.3.2",
    "detect-browser": "5.1.1",
    "element-closest-polyfill": "1.0.2",
    "fomantic-ui": "2.8.6",
    "moment": "2.24.0",
    "ngx-drag-drop": "2.0.0",
    "rxjs": "6.6.2",
    "tslib": "2.0.0",
    "typeface-roboto": "0.0.75",
    "uuid": "8.3.0",
    "zone.js": "0.10.3"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "0.1000.5",
    "@angular/cli": "10.0.5",
    "@angular/compiler-cli": "10.0.8",
    "@angular/language-service": "10.0.8",
    "@types/uuid": "8.0.1",
    "codelyzer": "6.0.0",
    "rimraf": "3.0.2",
    "rxjs-tslint-rules": "4.34.0",
    "ts-node": "8.10.2",
    "tslint": "6.1.3",
    "tslint-angular": "3.0.2",
    "typescript": "3.9.7",
    "webpack-bundle-analyzer": "3.8.0"
  }
}
```
</p>
</details>

## 🔬 Minimal Reproduction
I am actively trying to come up with a reproduction right now, but it's difficult because I'm just guessing at what things to try. This exact setup in a new app generated with `ng new` does work, the dev code is stripped.

## 🔥 Exception or Error
N/A

## 🌍 Your Environment
```
Angular CLI: 10.0.6
Node: 12.18.1
OS: win32 x64

Angular: 10.0.10
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.6
@angular-devkit/build-angular     0.1000.6
@angular-devkit/build-optimizer   0.1000.6
@angular-devkit/build-webpack     0.1000.6
@angular-devkit/core              10.0.6
@angular-devkit/schematics        10.0.6
@angular/cli                      10.0.6
@ngtools/webpack                  10.0.6
@schematics/angular               10.0.6
@schematics/update                0.1000.6
rxjs                              6.6.2
typescript                        3.9.7
webpack                           4.43.0
```

## Anything else relevant?
 * If it helps, something seems to be very wrong with optimization. [Here's a screenshot of our bundle stats.](https://user-images.githubusercontent.com/771134/91058484-81c35880-e5f6-11ea-81be-715fa7d6ff71.png) The largest `index.ts` there has [a parsed size of 373 KB](https://user-images.githubusercontent.com/771134/91063473-0bc1f000-e5fc-11ea-8599-dd8ca8a0ec0e.png), but it contains nothing other than interfaces, several enums, and several const enums. They are small enums and there are only a few, the vast majority of things in this barrel are interfaces (and none of the interfaces reference classes of any kind). This exact same project using Angular 9 instead of 10 makes that barrel 20 KB instead of 373 KB.
   * I think this is probably a separate issue because it's only a problem in Angular 10, whereas dev code isn't stripped in 9 or 10, but I'm mentioning it just in case.
 * Code bounded with `if (false) { }` is properly stripped.
 * Removing the services export from the end of `environment{.prod}.ts` does not fix the problem.
 * All CommonJS dependencies were removed, the CLI no longer complains about them, but the problem persists.

**Comment by amakhrov** [comment]
I see a similar issue in our project. I'm not sure if it only started after our migrating to Angular 10 though - haven't payed much attention before. We're currently on `10.0.3` version of angular-cli.

I was not able to make minimal reproduction.
What I've learned so far is that in our project this `environment` file is imported in a couple of places. And only one of them seems to have any effect on the issue: removing this import from that shared service resolves the issue.
I'll post here if I learn anything else.

**Comment by clydin** [comment]
Without a reproduction the definitive cause is hard to discern.  However, a potential cause is the use of the environment JS module (`environment.ts`/`environment.prod.ts`) in more than one generated output file.  This could be the case if the environment module is used in the main code and in the code for a lazy route.  When this happens, Webpack cannot concatenate the environment module with the main module (as happens in a new project) because the environment module needs to be accessible to two different output modules.  This then in turn prevents the optimizer from inlining the production property value since the environment object is now essentially an import from another module and not a local variable.

When this happens code similar to the following (which represents a separate Webpack module) should end up in the main output file for the application:
```
    AytR: function (module, __webpack_exports__, __webpack_require__) {
      "use strict";
      __webpack_require__.d(__webpack_exports__, "a", function () {
        return environment;
      });
      const environment = { production: !0 };
    },
```

**Comment by amakhrov** [comment]
I commented out all lazy-loaded modules / imports in our app to see if it helps - it didn't.
Running `ng build --prod` would create `vendor`, `runtime`, `polyfills`, and `main` chunks (each in 2 versions: es2015 and es5 for differential loading) - no more lazy loaded chunks. And still in `main` I see the environment module emitted as a module, not inlined. Which, as you pointed out, prevents the optimization.

Is it possible that referring to the `environment` file from multiple non-lazy modules (that end up in the same `main` chunk) still hass the same effect as with it being used across multiple chunks?

**Comment by clydin** [comment]
@amakhrov Yes, it is possible but generally less likely.  To further analyze the problem you can run a build with the verbose option and review the output using this command: `ng build --prod --verbose &> log.txt`.  You will want to search the log file for `ModuleConcatenation bailout`.  This should provide additional information on why Webpack did not concatenate the environment module within your application.

Also, a production application should typically not have a vendor chunk.  Splitting out the vendor code will result in a large size de-optimization.  The vendor chunk option is primarily intended for use during development to reduce the rebuild time in watch mode.

**Comment by amakhrov** [comment]
Thanks for the suggestion @clydin 

The log doesn't include any mention of `ModuleConcatenation`.
Also log seems to be truncated prematurely:
```
WARNING in Terser Plugin: Condition left of || always false [../node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:18378,19]

WARNING in Terser Plugin: Condition left of || always false [../node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:18380,23]

WARNING in
✨  Done in 241.75s.
```

I'm not sure if it has anything to do with the original issue. Anyway, I assume ModuleConcatenation should execute before TerserPlugin, so the log truncated at some Terser output should not have hidden any logs from ModuleConcatenation, right?

I totally realize there is no much you can do without exact repro steps. I'll post here if I manage to get it.

In the meantime, I wanted to also discuss this:
> This could be the case if the environment module is used in the main code and in the code for a lazy route

Does it mean that even if I manage to fix the issue in our project right now, it can still return later if we try to consume `environment` file from lazy routes? Is there any workaround for that?

**Comment by amakhrov** [comment]
Aha! Turned out, generated `stats.json` also includes optimization bailouts details!
So indeed for environment.ts there was an bailout record:
```
"ModuleConcatenation bailout: Cannot concat with ./environments/environment.ts because of ./app/shared/services/api-client/index.ts",
```

Then looking further I see the bailout for that `api-client/index.ts` module:
```
"ModuleConcatenation bailout: Module exports are unknown"
```

This seems to be it! In `api-client/index.ts` **we have a wildcard export**
```
export * from "./models";
```
(those models contain definitions of numerous interfaces and enums, autogenerated from a swagger spec).

So a workaround was to split `environment.ts` into 2 files, making sure the one with the `production` flag is not directly or indirectly imported from that api-client.

Funny enough, it still didn't solve our root issue: ngrx/store-devtools are still included in the bundle, even though now truly unused. There is a separate bailout message for this one:
```
ModuleConcatenation bailout: Cannot concat with ../node_modules/@ngrx/store-devtools/fesm2015/store-devtools.js (<- Module uses eval())"
```
which probably prevents tree shaking as well.

Anyway, thanks @clydin for pointing me out to investigating of ModuleConcatenation bailouts. I'm sure it's gonna help me a lot in future also.

**Comment by vaindil** [comment]
Thanks to both of you for your replies, they've been incredibly helpful. I just want to ask this question again:

> In the meantime, I wanted to also discuss this:
> 
>  >   This could be the case if the environment module is used in the main code and in the code for a lazy route
> 
> Does it mean that even if I manage to fix the issue in our project right now, it can still return later if we try to consume environment file from lazy routes? Is there any workaround for that?

**Comment by alan-agius4** [comment]
Hi @vaindil, 

To answer your question, indeed if you use the environment configuration in multiple chunks, Webpack will not be able to concatenate the module. From the screenshot provided I also see that you are using barrel files, which typically are one of the main reasons for optimisation bailouts and therefore dead code retention. At the moment, if you are using the environments file in multiple chunks there isn't much you can do. part from splitting the environments file per chunk.

I do recommend periodic analyses and auditing of the bundles to verify that there aren't any de-optimization happening with newly introduced code.

**Comment by vaindil** [comment]
@alan-agius4 Thanks for your answer, that's really unfortunate. I'm going to open a separate issue for the bundle size issue with that screenshot, because Angular 10 alone increased our bundle size by 1MB. Need to do some more analysis though to troubleshoot first. Thanks!
